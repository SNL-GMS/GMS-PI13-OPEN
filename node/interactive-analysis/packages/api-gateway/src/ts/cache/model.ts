import Immutable from 'immutable';
import { Association, TimeRange } from '../common/model';
import { Event } from '../event/model-and-schema/model';
import { QcMask } from '../qc-mask/model';
import { SignalDetection } from '../signal-detection/model';
import { UserProfile } from '../user-profile/model';

/**
 * Specifies the type (method) for committing to the Global Cache.
 */
export type CommitFunc<T> = (items: T[]) => void;

/**
 * User Cache Processor interface.
 * Defines all of the common method definitions required
 * for a user cache item.
 */
// tslint:disable-next-line: interface-name
export interface IUserCacheItem<T> {
  has(id: string): boolean;
  get(id: string): T;
  getAll(): T[];
  set(item: T): void;
  setAll(items: T[]): void;
  remove(item: T): void;
  removeAll(item: T[]): void;
  commitAll(): void;
  commitWithIds(ids: string[]): void;
  updateFromGlobalCache(cachedEntries: T[], overwrite: boolean): void;
}

/**
 * User Cache interface.
 * Defines all of the common method definitions required for a user cache.
 */
// tslint:disable-next-line: interface-name
export interface IUserCache {
  getTimeRange(): TimeRange;
  setTimeRange(timeRange: TimeRange): void;
  getUserProfile(): UserProfile;
  setUserProfile(userProfile: UserProfile): void;
  getOpenEventId(): string;
  setOpenEventId(id: string): void;
  getEventById(id: string): Event | undefined;
  getEvents(): Event[];
  getInvalidEventIds(): string[];
  setEvent(userAction: UserActionDescription, event: Event): void;
  setEvents(userAction: UserActionDescription, events: Event[]): void;
  updateEventsFromGlobalCache(events: Event[], overwrite: boolean): void;
  commitAllEvents(): void;
  commitEventsWithIds(ids: string[]): void;
  getSignalDetectionById(id: string): SignalDetection | undefined;
  getSignalDetections(): SignalDetection[];
  getInvalidSignalDetectionIds(): string[];
  setSignalDetection(userAction: UserActionDescription, sd: SignalDetection): void;
  setSignalDetections(userAction: UserActionDescription, sds: SignalDetection[]): void;
  updateSignalDetectionsFromGlobalCache(sds: SignalDetection[], overwrite: boolean): void;
  commitAllSignalDetections(): void;
  commitSignalDetectionsWithIds(ids: string[]): void;
  getAssociations(): Association[];
  getSignalDetectionAndEventBySdId(id): { signalDetection: SignalDetection; event: Event };
  setEventsAndSignalDetections(
    userAction: UserActionDescription,
    events: Event[],
    sds: SignalDetection[]
  ): void;
  getRedoPriorityOrder(id: string): number | undefined;
  getHistory(): History[];
  undoHistory(numberOfItems: number): UserCacheData;
  redoHistory(numberOfItems: number): UserCacheData;
  undoHistoryById(id: string): UserCacheData;
  redoHistoryById(id: string): UserCacheData;
  getEventHistory(id: string): History[];
  undoEventHistory(numberOfItems: number): UserCacheData;
  redoEventHistory(numberOfItems: number): UserCacheData;
  undoEventHistoryById(id: string): UserCacheData;
  redoEventHistoryById(id: string): UserCacheData;
  clearHistory(): void;
}

/**
 * User context for apollo and user sessions
 */
export interface UserContext {
  readonly sessionId: string;
  readonly userName: string;
  readonly userCache: IUserCache;
  readonly userRole: string;
}

/**
 * Represents the user cache data from the user cache.
 */
export interface UserCacheData {
  readonly events: Event[];
  readonly signalDetections: SignalDetection[];
}

/**
 * Description of a user action preformed
 */
export enum UserActionDescription {
  UNKNOWN = 'Unknown',
  CHANGE_SIGNAL_DETECTION_ASSOCIATIONS_ASSOCIATE = 'Associate',
  CHANGE_SIGNAL_DETECTION_ASSOCIATIONS_ASSOCIATE_MULTIPLE = 'Associate multiple detections',
  CHANGE_SIGNAL_DETECTION_ASSOCIATIONS_UNASSOCIATE = 'Unassociate',
  CHANGE_SIGNAL_DETECTION_ASSOCIATIONS_UNASSOCIATE_MULTIPLE = 'Unassociate multiple detections',

  // signal detection actions
  CREATE_DETECTION = 'Create detection',
  REJECT_DETECTION = 'Reject detection',
  REJECT_MULTIPLE_DETECTIONS = 'Reject multiple detections',
  UPDATE_DETECTION_RE_TIME = 'Time',
  UPDATE_DETECTION_RE_PHASE = 'Phase',
  UPDATE_MULTIPLE_DETECTIONS_RE_PHASE = 'Phase multiple detections',
  UPDATE_DETECTION_AMPLITUDE = 'Update detection amplitude',
  UPDATE_DETECTION_REVIEW_AMPLITUDE = 'Update detection review amplitude',
  UPDATE_DETECTION = 'Update detection',
  COMPUTE_FK = 'Fk',
  COMPUTE_MULTIPLE_FK = 'Fk multiple detections',

  // event actions
  CREATE_EVENT = 'Create event',
  UPDATE_EVENT_LOCATE = 'Locate event',
  UPDATE_EVENT_FROM_SIGNAL_DETECTION_CHANGE = 'Update event from signal detection change',
  UPDATE_EVENT_FEATURE_PREDICTIONS = 'Update event feature predictions',
  UPDATE_EVENT_STATUS_OPEN_FOR_REFINEMENT = 'Event opened for refinement',
  UPDATE_EVENT_MARK_COMPLETE = 'Event marked complete',
  UPDATE_EVENT_PREFERRED_HYP = 'Update event preferred hypothesis',
  UPDATE_EVENT_MAGNITUDE = 'Change Magnitude defining settings',
  SAVE_EVENT = 'Save event'
}

/**
 * Description of a user action performed that cannot be undone and will not be in the history
 */
export const NON_UNDO_ACTIONS: UserActionDescription[] = [
  UserActionDescription.UPDATE_EVENT_FEATURE_PREDICTIONS,
  UserActionDescription.UPDATE_EVENT_STATUS_OPEN_FOR_REFINEMENT,
  UserActionDescription.UPDATE_EVENT_MARK_COMPLETE,
  UserActionDescription.SAVE_EVENT,
  UserActionDescription.UPDATE_DETECTION_REVIEW_AMPLITUDE
];

/**
 * Defines the interface of User Action.
 */
export interface UserAction {
  /**
   * The user action description
   */
  readonly description: UserActionDescription;

  /**
   * Returns the toString value of a user action.
   *
   * @returns the string representation of the action
   */
  toString(): string;
}

/**
 * Defines the hypothesis type for a hypothesis change.
 */
export enum HypothesisType {
  EventHypothesis = 'EventHypothesis',
  SignalDetectionHypothesis = 'SignalDetectionHypothesis'
}

/**
 * Represents hypothesis information for a change.
 * Defines a mapping from a hypothesis id to the main
 * object id.
 */
export interface HypothesisChangeInformation {
  readonly id: string;
  readonly hypothesisId: string;
  readonly type: HypothesisType;
  readonly parentId: string;
  readonly userAction: UserAction;
}

/**
 * Represents a history change.
 */
export interface HistoryChange {
  readonly id: string;
  readonly active: boolean;
  readonly eventId: string;
  readonly conflictCreated: boolean;
  readonly hypothesisChangeInformation: HypothesisChangeInformation;
}

/**
 * Represents the history.
 */
export interface History {
  readonly id: string;
  readonly description: UserActionDescription;
  readonly changes: Immutable.List<HistoryChange>;
}

/**
 * Represents data that has been created or updated and needs to
 * be sent to the UI Clients
 *
 * Currently only include events and sds
 */
export interface DataPayload {
  readonly events: Event[];
  readonly sds: SignalDetection[];
  readonly qcMasks: QcMask[];
}

/**
 * Invalid data; typically data that has been deleted.
 */
export interface InvalidData {
  readonly eventIds: string[];
  readonly signalDetectionIds: string[];
}
