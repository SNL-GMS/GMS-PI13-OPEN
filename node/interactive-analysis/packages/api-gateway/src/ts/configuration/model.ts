import Immutable from 'immutable';
import { WaveformFilterDefinition } from '../waveform-filter/model';

/**
 * Interface for analyst configuration used to determine fields based on roles
 */
export interface AnalystConfiguration {
  readonly defaultNetwork: string;
  readonly defaultFilters: WaveformFilterDefinition[];
}

/**
 * Soh specific configuration
 */
export interface SohConfiguration {
  stationSohControlConfiguration: {
    reprocessingPeriod: string;
    displayedStationGroups: string[];
    rollupStationSohTimeTolerance: string;
  };
  stationSohMonitoringDisplayParameters: {
    redisplayPeriod: string;
    acknowledgementQuietDuration: string;
    availableQuietDurations: string[];
    sohStationStaleDuration: string;
    sohHistoricalDurations: string[];
  };
}

/**
 * Common configuration
 */
export interface CommonConfiguration {
  systemMessageLimit: number;
}

/**
 * Overall configuration
 */
export interface Configuration {
  /**
   * AnalystConfiguration keyed on user role
   */
  analystConfiguration: Immutable.Map<string, AnalystConfiguration>;

  /**
   * SohConfigurationOld
   */
  sohConfiguration: SohConfiguration;

  /**
   * CommonConfiguration keyed on user role
   */
  commonConfiguration: Immutable.Map<string, CommonConfiguration>;
}

/**
 * Selector interface for config service
 */
export interface Selector {
  criterion: string;
  value: string;
}

/**
 * Analyst configurations loaded from service
 */
export enum AnalystConfigs {
  DEFAULT = 'ui.analyst-settings'
}

/**
 * SOH configurations loaded from service
 */
export const SohConfig = 'ui.soh-settings';

/**
 * Common configurations loaded from service
 */
export enum CommonConfigs {
  DEFAULT = 'ui.common-settings'
}

/**
 * Default user role
 */
export const defaultUserRole = 'DEFAULT';
