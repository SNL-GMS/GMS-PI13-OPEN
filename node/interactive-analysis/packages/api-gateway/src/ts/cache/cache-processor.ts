import Immutable from 'immutable';
import cloneDeep from 'lodash/cloneDeep';
import { ChannelSegment, TimeSeries } from '../channel-segment/model';
import { WorkspaceState } from '../common/model';
import { Event } from '../event/model-and-schema/model';
import { gatewayLogger } from '../log/gateway-logger';
import { QcMask } from '../qc-mask/model';
import { SignalDetection } from '../signal-detection/model';
import { ProcessingStationData } from '../station/processing-station/model';
import { ReferenceStationData } from '../station/reference-station/model';
import { getInitialTimeRange } from '../util/workflow-util';
import { WaveformFilterDefinition } from '../waveform-filter/model';
import { ProcessingActivityInterval, WorkflowData } from '../workflow/model';
import { GlobalCache } from './global-cache';
import { UserCache } from './user-cache';
import { UserCacheItem } from './user-cache-item';

/**
 * API gateway Cache Processor for a single global data cache that represents the
 * OSD state for all users.
 */
export class CacheProcessor {
  /** The singleton instance */
  private static instance: CacheProcessor;

  /**
   * Shared workspace state
   */
  private workspaceState: WorkspaceState;

  /** The singleton instance of Global Cache */
  private globalCache: GlobalCache;

  /**
   * Stored User Cache
   */
  private userCaches: Immutable.Map<string, UserCache>;

  /**
   * Constructor
   */
  private constructor() {
    /* no-op */
  }

  /**
   * Returns the singleton instance of the cache processor.
   * @returns the instance of the cache processor
   */
  public static Instance(): CacheProcessor {
    if (CacheProcessor.instance === undefined) {
      CacheProcessor.instance = new CacheProcessor();
      CacheProcessor.instance.initialize();
    }
    return CacheProcessor.instance;
  }

  // ----- User Cache ------

  /**
   * Gets or creates a user cache for the given id
   * @param id the user session id
   * @returns a user cache object
   */
  public getCacheForUser(id: string): UserCache {
    if (id === undefined) {
      gatewayLogger.error(
        'Undefined session id, likely caused by an async resolver that is not reading the session context'
      );
    }
    if (this.userCaches.has(id)) {
      return this.userCaches.get(id);
    }
    const eventCache = new UserCacheItem<Event>(
      // clone the events to ensure that the global cache is not accidentally modified, consider freezing
      cloneDeep(CacheProcessor.Instance().globalCache.getEvents()),
      (events: Event[]) => CacheProcessor.Instance().commitEvents(events)
    );

    const sdCache = new UserCacheItem<SignalDetection>(
      // clone the events to ensure that the global cache is not accidentally modified, consider freezing
      cloneDeep(CacheProcessor.Instance().globalCache.getSignalDetections()),
      (sds: SignalDetection[]) => CacheProcessor.Instance().commitSignalDetections(sds)
    );

    const newUserCache = new UserCache(getInitialTimeRange(), undefined, eventCache, sdCache);
    this.userCaches = this.userCaches.set(id, newUserCache);
    gatewayLogger.info(`Creating new User Cache with ID: ${id}`);
    return newUserCache;
  }

  /**
   * Deletes the user caches for a session ID
   * @param sessionId the unique session id
   */
  public deleteUserCache(sessionId: string): void {
    if (this.userCaches.has(sessionId)) {
      this.userCaches = this.userCaches.delete(sessionId);
    }
  }

  // ----- Workspace State Functions ------

  /**
   * Returns the workspace state
   * @returns the workspace state
   */
  public getWorkspaceState(): WorkspaceState {
    return cloneDeep(this.workspaceState);
  }

  /**
   * Adds or updates event to user map in workspace state
   * @param eventId event id to remove user from
   * @param userName the username to remove
   */
  public addOrUpdateEventToUser(eventId: string, userName: string): void {
    const eventToUsers = this.workspaceState.eventToUsers.find(
      evToUsers => evToUsers.eventId === eventId
    );
    if (eventToUsers) {
      // Event is already in the map so if the username is not in the list push it
      if (eventToUsers.userNames.indexOf(userName) < 0) {
        eventToUsers.userNames.push(userName);
      }
    } else {
      // Event is not in map, create new username array
      this.workspaceState.eventToUsers.push({ eventId, userNames: [userName] });
    }
  }

  /**
   * Removes a user from the event in the workspace state
   * @param eventId event id to remove user from
   * @param userName the username to remove
   */
  public removeUserFromEvent(eventId: string, userName: string): void {
    const eventToUsers = this.workspaceState.eventToUsers.find(
      evToUsers => evToUsers.eventId === eventId
    );
    const index = eventToUsers ? eventToUsers.userNames.indexOf(userName) : undefined;
    if (index >= 0) {
      eventToUsers.userNames.splice(index, 1);
    }
  }

  // ----- Configuration Functions ------

  /**
   * Gets the configuration from the global cache
   * @returns the configuration from the global cache
   */
  public getConfiguration(): Immutable.Map<string, string> {
    return this.globalCache.getConfiguration();
  }

  /**
   * Set the configuration in the global cache
   * @param configuration the configuration
   */
  public setConfiguration(configuration: Immutable.Map<string, string>): void {
    this.globalCache.setConfiguration(configuration);
  }

  // ----- Reference Station Data Functions ------

  /**
   * Returns true if the station data exists; false otherwise
   * @param id the unique id of the station
   * @returns true if the station data exists; false otherwise
   */
  public hasReferenceStationData(id: string): boolean {
    return this.globalCache.hasReferenceStationData(id);
  }

  /**
   * Gets the station data from the global cache for the given id
   * @param name the unique id (network name)
   * @returns the station data
   */
  public getReferenceStationData(name: string): ReferenceStationData {
    return this.globalCache.getReferenceStationData(name);
  }

  /**
   * Gets the station data from the global cache
   * @returns the station data
   */
  public getReferenceAllStationData(): Immutable.Map<string, ReferenceStationData> {
    return this.globalCache.getAllReferenceStationData();
  }

  /**
   * Set the station data in the global cache
   * @param stationData the station data to save/update
   */
  public setReferenceStationData(stationData: ReferenceStationData): void {
    this.globalCache.setReferenceStationData(stationData);
  }

  // ----- Processing Station Data Functions ------

  /**
   * Gets the station data from the global cache for the given id
   * @param name the unique id (network name)
   * @returns the station data
   */
  public getProcessingStationData(): ProcessingStationData {
    return this.globalCache.getProcessingStationData();
  }

  // ----- Workflow Data Functions ------

  /**
   * Returns the workflow data from the global cache.
   *
   * @returns the workflow data
   */
  public getWorkflowData(): WorkflowData {
    return this.globalCache.getWorkflowData();
  }

  /**
   * Sets the workflow data in the global cache.
   *
   * @param workflow the workflow data to set
   */
  public setWorkflowData(workflow: WorkflowData): void {
    this.globalCache.setWorkflowData(workflow);
  }

  /**
   * Returns the current open activity from the global cache.
   *
   * @returns the current open activity.
   */
  public getCurrentOpenActivity(): ProcessingActivityInterval {
    return this.globalCache.getCurrentOpenActivity();
  }

  /**
   * Sets the current open activity in the global cache.
   *
   * @param activity the activity to set
   */
  public setCurrentOpenActivity(activity: ProcessingActivityInterval): void {
    this.globalCache.setCurrentOpenActivity(activity);
  }

  // ----- Waveform Filter Definition Functions ------

  /**
   * Gets the waveform filter definitions from the global cache
   */
  public getWaveformFilterDefinitions(): Immutable.List<WaveformFilterDefinition> {
    return this.globalCache.getWaveformFilterDefinitions();
  }

  /**
   * Set the waveform filter definitions in the global cache
   * @param definitions the waveform filter definitions to save/update
   */
  public setWaveformFilterDefinitions(
    definitions: Immutable.List<WaveformFilterDefinition> | WaveformFilterDefinition[]
  ): void {
    this.globalCache.setWaveformFilterDefinitions(definitions);
  }

  // ----- Channel Segment Functions ------

  /**
   * Returns true if the channel segment exists; false otherwise
   * @param id the unique channel segment id
   * @returns true if the channel segment exists; false otherwise
   */
  public hasChannelSegment(id: string): boolean {
    return this.globalCache.hasChannelSegment(id);
  }

  /**
   * Gets the channel segment from the global cache for the given id
   * @param id the unique id
   * @returns the channel segment
   */
  public getChannelSegment(id: string): ChannelSegment<TimeSeries> {
    return this.globalCache.getChannelSegment(id);
  }

  /**
   * Gets the channel segments from the global cache
   * @returns the channel segments
   */
  public getChannelSegments(): Immutable.Map<string, ChannelSegment<TimeSeries>> {
    return this.globalCache.getChannelSegments();
  }

  /**
   * Set the channel segments in the global cache
   * @param channelSegment the channel segments to save/update
   */
  public setChannelSegments(channelSegments: ChannelSegment<TimeSeries>[]): void {
    this.globalCache.setChannelSegments(channelSegments);
  }

  /**
   * Set the channel segment in the global cache
   * @param channelSegment the channel segment to save/update
   */
  public setChannelSegment(channelSegment: ChannelSegment<TimeSeries>): void {
    this.globalCache.setChannelSegment(channelSegment);
  }

  // ----- Event Functions ------

  /**
   * When a component adds new data (queried from the OSD),
   * update the Global Cache and sync to the individual
   * user caches (cloned global caches);
   * @param events the events to be added to the global cache
   */
  public addLoadedEventsToGlobalCache(events: Event[]): void {
    // global cache add all of them
    events.forEach(event => {
      this.globalCache.setEvent(event);
    });
    this.pushAndPublishUpdatedGlobalEvents(events.map(event => event.id));
  }

  // ----- Signal Detection Functions ------

  /**
   * When a component adds new data (queried from the OSD),
   * update the Global Cache and sync to the individual
   * user caches (cloned global caches);
   * @param sds the signal detections to be added to the global cache
   */
  public addLoadedSdsToGlobalCache(sds: SignalDetection[]): void {
    // global cache add all of them
    sds.forEach(sd => {
      this.globalCache.setSignalDetection(sd);
    });
    this.pushAndPublishUpdatedGlobalSignalDetections(sds.map(sd => sd.id));
  }

  // ----- QC Mask Functions ------

  /**
   * Returns true if the qc mask exists; false otherwise
   * @param id the qc mask id
   * @returns true if exists; false otherwise
   */
  public hasQcMask(id: string): boolean {
    return this.globalCache.hasQcMask(id);
  }

  /**
   * Gets the qc mask from the global cache for the given id
   * @param id the qc mask id
   * @returns the qc mask
   */
  public getQcMask(id: string): QcMask {
    return this.globalCache.getQcMask(id);
  }

  /**
   * Gets the qc mask from the global cache
   * @returns the qc masks
   */
  public getQcMasks(): Immutable.Map<string, QcMask> {
    return this.globalCache.getQcMasks();
  }

  /**
   * Set the qc masks in the global cache
   * @param qcMasks the qc masks to save/update
   */
  public setQcMasks(qcMasks: QcMask[]): void {
    this.globalCache.setQcMasks(qcMasks);
  }

  /**
   * Set the qc mask in the global cache
   * @param qcMask the qc mask to save/update
   */
  public setQcMask(qcMask: QcMask): void {
    this.globalCache.setQcMask(qcMask);
  }

  // ----- Private Functions ------

  /**
   * Initialize the Cache Processor (global cache and user caches)
   */
  private initialize(): void {
    this.globalCache = new GlobalCache();
    this.userCaches = Immutable.Map<string, UserCache>();

    this.workspaceState = {
      eventToUsers: []
    };
  }

  /**
   * Commit signal detections to the global cache.
   * Means that a user has saved (committed).
   * @param signalDetections the signal detections to commit
   */
  private commitSignalDetections(signalDetections: SignalDetection[]): void {
    // Walk thru all the user cache SD entries and set them on Global cache
    // then clear user cache list
    signalDetections.forEach(uSd => {
      this.globalCache.setSignalDetection(uSd);
    });
    this.pushAndPublishUpdatedGlobalSignalDetections(
      signalDetections.map(userSd => userSd.id),
      true
    );
  }

  /**
   * Push and publish updated global cache events to the user caches.
   * @param ids event ids to publish
   * @param overwrite if true, the user cache will be overwritten with new data
   */
  private pushAndPublishUpdatedGlobalSignalDetections(
    ids: string[],
    overwrite: boolean = false
  ): void {
    this.userCaches.forEach(userCache => {
      userCache.updateSignalDetectionsFromGlobalCache(
        [
          ...cloneDeep(
            this.globalCache
              .getSignalDetections()
              .filter((value, key) => ids.find(id => id === key))
              .values()
          )
        ],
        overwrite
      );
    });
  }

  /**
   * Commit events to the global cache.
   * Means that a user has saved (committed).
   * @param signalDetections the events to commit
   */
  private commitEvents(userEvents: Event[]): void {
    // Walk thru all the user cache event entries and set them on Global cache
    // then clear user cache list
    userEvents.forEach(uEvent => {
      this.globalCache.setEvent(uEvent);
    });
    this.pushAndPublishUpdatedGlobalEvents(
      userEvents.map(event => event.id),
      true
    );
  }

  /**
   * Push and publish updated global cache events to the user caches.
   * @param ids event ids to publish
   * @param overwrite if true, the user cache will be overwritten with new data
   */
  private pushAndPublishUpdatedGlobalEvents(ids: string[], overwrite: boolean = false): void {
    this.userCaches.forEach(userCache => {
      userCache.updateEventsFromGlobalCache(
        [
          ...cloneDeep(
            this.globalCache
              .getEvents()
              .filter((value, key) => ids.find(id => id === key))
              .values()
          )
        ],
        overwrite
      );
    });
  }
}
