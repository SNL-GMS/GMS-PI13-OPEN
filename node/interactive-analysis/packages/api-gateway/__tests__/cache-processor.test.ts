import { CacheProcessor } from '../src/ts/cache/cache-processor';
import { UserActionDescription } from '../src/ts/cache/model';
import { Event, EventStatus } from '../src/ts/event/model-and-schema/model';
import { SignalDetection } from '../src/ts/signal-detection/model';

const contextBob = { sessionId: 'test1', userName: 'Uncle Bob' };
const contextMaria = { sessionId: 'test2', userName: 'Aunt Maria' };
const contextLetters = { sessionId: 'test3', userName: 'Letters' };
const contextRasco = { sessionId: 'test4', userName: 'Rasco' };

const mockSdBob: SignalDetection = {
  id: 'bob',
  monitoringOrganization: 'bob',
  stationName: 'bob',
  signalDetectionHypotheses: [],
  currentHypothesis: undefined,
  associations: [],
  hasConflict: false
};

const mockSdMaria: SignalDetection = {
  id: 'maria',
  monitoringOrganization: 'maria',
  stationName: 'maria',
  signalDetectionHypotheses: [],
  currentHypothesis: undefined,
  associations: [],
  hasConflict: false
};

const mockEventBob: Partial<Event> = {
  id: 'bob',
  monitoringOrganization: 'bob',
  preferredEventHypothesisHistory: [],
  finalEventHypothesisHistory: [],
  hypotheses: [],
  status: EventStatus.OpenForRefinement,
  currentEventHypothesis: {
    eventHypothesis: undefined,
    processingStageId: 'bob'
  },
  associations: [],
  signalDetectionIds: [],
  hasConflict: false
};

const mockEventMaria: Partial<Event> = {
  id: 'maria',
  monitoringOrganization: 'maria',
  preferredEventHypothesisHistory: [],
  finalEventHypothesisHistory: [],
  hypotheses: [],
  status: EventStatus.OpenForRefinement,
  currentEventHypothesis: {
    eventHypothesis: undefined,
    processingStageId: 'maria'
  },
  associations: [],
  signalDetectionIds: [],
  hasConflict: false
};

const cacheBob = CacheProcessor.Instance().getCacheForUser(contextBob.sessionId);
const cacheMaria = CacheProcessor.Instance().getCacheForUser(contextMaria.sessionId);

const cacheRascoEvent: Partial<Event> = {
  id: 'rasco',
  monitoringOrganization: 'rasco',
  preferredEventHypothesisHistory: [],
  finalEventHypothesisHistory: [],
  hypotheses: [],
  status: EventStatus.OpenForRefinement,
  currentEventHypothesis: {
    eventHypothesis: undefined,
    processingStageId: 'rasco'
  },
  associations: [],
  signalDetectionIds: [],
  hasConflict: false
};

const sharedEvent: Partial<Event> = {
  id: 'shared-event-1',
  monitoringOrganization: 'shared-event-1',
  preferredEventHypothesisHistory: [],
  finalEventHypothesisHistory: [],
  hypotheses: [],
  status: EventStatus.OpenForRefinement,
  currentEventHypothesis: {
    eventHypothesis: undefined,
    processingStageId: 'shared-event-1'
  },
  associations: [],
  signalDetectionIds: [],
  hasConflict: false
};

const sharedSd: Partial<SignalDetection> = {
  id: 'shared-sd-1',
  monitoringOrganization: 'shared-sd-1',
  stationName: 'shared-sd-1',
  signalDetectionHypotheses: [],
  currentHypothesis: undefined,
  associations: [],
  hasConflict: false
};

CacheProcessor.Instance().addLoadedEventsToGlobalCache([sharedEvent as any]);
CacheProcessor.Instance().addLoadedSdsToGlobalCache([sharedSd as any]);

const cacheLetters = CacheProcessor.Instance().getCacheForUser(contextLetters.sessionId);
const cacheRasco = CacheProcessor.Instance().getCacheForUser(contextRasco.sessionId);

xdescribe('User Cache Access', () => {
  it('A users local cache is not affects by unsaved changes to another cache', () => {
    cacheBob.setSignalDetection(UserActionDescription.UPDATE_DETECTION, mockSdBob);
    cacheMaria.setSignalDetection(UserActionDescription.UPDATE_DETECTION, mockSdMaria);

    expect(cacheBob.getSignalDetectionById('a').stationName).toEqual('bob');
  });
  it('After a push, the other users cache is updated', () => {
    cacheBob.setEvent(UserActionDescription.UNKNOWN, mockEventBob as any);
    cacheMaria.setEvent(UserActionDescription.UNKNOWN, mockEventMaria as any);
    cacheMaria.commitAllEvents();

    expect(cacheBob.getEventById('a').monitoringOrganization).toEqual('maria');
  });
});

xdescribe('Publishing individual changes', () => {
  it('Saving one event does not override the users other events', () => {
    const changedSharedEvent: any = {
      id: 'sharedEvent',
      monitoringOrganization: 'changed'
    };
    const changedSharedEventTwo: any = {
      id: 'sharedEventTwo',
      monitoringOrganization: 'changed'
    };
    cacheLetters.setEvent(UserActionDescription.UNKNOWN, changedSharedEvent);
    cacheLetters.setEvent(UserActionDescription.UNKNOWN, changedSharedEventTwo);
    cacheLetters.commitEventsWithIds(['sharedEvent']);
    expect(cacheLetters.getEventById('sharedEventTwo').monitoringOrganization).toEqual('changed');
  });
  it('Saving one event correct populates the other user', () => {
    cacheRasco.setEvent(UserActionDescription.UNKNOWN, cacheRascoEvent as any);
    cacheRasco.commitEventsWithIds([cacheRascoEvent.id]);
    expect(cacheLetters.getEventById(cacheRascoEvent.id).monitoringOrganization).toEqual(
      cacheRasco.getEventById(cacheRascoEvent.id).monitoringOrganization
    );
  });
});

describe('Workspace State Test', () => {
  test('Adding event to users to workspace state', () => {
    let workspaceState = CacheProcessor.Instance().getWorkspaceState();
    expect(workspaceState).toBeDefined();
    expect(workspaceState).toMatchSnapshot();

    CacheProcessor.Instance().addOrUpdateEventToUser('test', 'user1');
    workspaceState = CacheProcessor.Instance().getWorkspaceState();
    expect(workspaceState).toBeDefined();
    expect(workspaceState).toMatchSnapshot();

    CacheProcessor.Instance().removeUserFromEvent('test', 'user1');
    workspaceState = CacheProcessor.Instance().getWorkspaceState();
    expect(workspaceState).toBeDefined();
    expect(workspaceState).toMatchSnapshot();
  });
});
