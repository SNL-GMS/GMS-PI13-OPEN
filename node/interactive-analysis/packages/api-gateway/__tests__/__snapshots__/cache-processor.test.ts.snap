// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Workspace State Test Adding event to users to workspace state 1`] = `
Object {
  "eventToUsers": Array [],
}
`;

exports[`Workspace State Test Adding event to users to workspace state 2`] = `
Object {
  "eventToUsers": Array [
    Object {
      "eventId": "test",
      "userNames": Array [
        "user1",
      ],
    },
  ],
}
`;

exports[`Workspace State Test Adding event to users to workspace state 3`] = `
Object {
  "eventToUsers": Array [
    Object {
      "eventId": "test",
      "userNames": Array [],
    },
  ],
}
`;
